File: 146-lru-cache.java
class LRUCache {
    private int capacity;
    private int count;
    private DLinkedNode head;
    private DLinkedNode tail;
    private Map<Integer, DLinkedNode> cache = new HashMap<>();
    
    class DLinkedNode {
        int key;
        int value;
        DLinkedNode prev;
        DLinkedNode next;
        
        DLinkedNode(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }
    
    private void insertAfterNode(DLinkedNode node, DLinkedNode newNode) {
        DLinkedNode next = node.next;
        
        node.next = newNode;
        newNode.next = next;
        
        newNode.prev = node;    
        next.prev = newNode;      
    }
    
    private void removeNode(DLinkedNode node) {
        DLinkedNode prev = node.prev;
        DLinkedNode next = node.next;
        prev.next = next;
        next.prev = prev;
    }
    
    private void moveToHead(DLinkedNode node) {
        removeNode(node);
        insertAfterNode(head, node);
    }
    
    private DLinkedNode removePreviousOfTail() {
        DLinkedNode node = tail.prev;
        removeNode(node);
        
        return node;
    }
    
    private void intializeHeadTail() {
        head = new DLinkedNode(-1, -1);
        tail = new DLinkedNode(-1, -1);
        head.next = tail;
        tail.prev = head;
    }
    
    public LRUCache(int capacity) {
        this.capacity = capacity;
        intializeHeadTail();
    }
    
    public int get(int key) {
        if(!cache.containsKey(key)) { return -1; }
        
        DLinkedNode node = cache.get(key);
        moveToHead(node);
        
        return node.value;
    }
    
    public void put(int key, int value) {
        if(cache.containsKey(key)) {
            DLinkedNode node = cache.get(key);
            node.value = value;
            moveToHead(node);
        }
        else {
            count++;
            
            DLinkedNode createdNode = new DLinkedNode(key, value);
            cache.put(key, createdNode);
            insertAfterNode(head, createdNode);
            
            if(count > capacity) {
                DLinkedNode tail = this.removePreviousOfTail();
                cache.remove(tail.key);
                count--;
            }
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 146_LRU_Cache.py
I implement LRU Cache using two structures: a hash map and a doubly linked list.
The hash map gives O(1) access to nodes.
The linked list keeps keys ordered by recency—most recent at the head, least recent at the tail.
Whenever I get or update a key, I move its node to the head.
When inserting a new key and the cache is full, I evict the node at the tail, which is the least recently used.
This design guarantees O(1) get and put.

class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None


class LRUCache:

    def __init__(self, capacity):
        self.capacity = capacity
        self.cache = {}  # key -> Node

        # Dummy head and tail to avoid edge cases
        self.head = Node(-1, -1)
        self.tail = Node(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head

    # -------------------------------------------------

    def get(self, key):
        if key not in self.cache:
            return -1

        node = self.cache[key]
        self._remove(node)
        self._add_to_front(node)
        return node.value

    # -------------------------------------------------

    def put(self, key, value):
        if key in self.cache:
            # Update existing node
            node = self.cache[key]
            node.value = value
            self._remove(node)
            self._add_to_front(node)
            return

        # Evict LRU if full
        if len(self.cache) == self.capacity:
            lru = self.tail.prev
            self._remove(lru)
            del self.cache[lru.key]

        # Insert new node at front
        new_node = Node(key, value)
        self._add_to_front(new_node)
        self.cache[key] = new_node

    # -------------------------------------------------

    def _remove(self, node):
        """Detach a node from the doubly linked list."""
        prev_node = node.prev
        next_node = node.next
        prev_node.next = next_node
        next_node.prev = prev_node

    def _add_to_front(self, node):
        """Insert a node right after head (most recent)."""
        first = self.head.next
        self.head.next = node
        node.prev = self.head
        node.next = first
        first.prev = node
